libraries/llvm/primitives.c

```diff
@@ -1,21 +1,62 @@
 #ifndef EFFEKT_PRIMITIVES_C
 #define EFFEKT_PRIMITIVES_C
 
-#include <inttypes.h>
 
+// types
 
 struct Pos {
-    uint64_t tag; // local tag
+    uint64_t tag; // type-local tag
     void *obj; // heap object
 };
 
+typedef int64_t Int;
+typedef double Double;
 
-void c_println_int64(const int64_t n) {
+static const struct Pos Unit = (struct Pos) { .tag = 0, .obj = NULL, };
+static const struct Pos BooleanFalse = (struct Pos) { .tag = 0, .obj = NULL, };
+static const struct Pos BooleanTrue = (struct Pos) { .tag = 1, .obj = NULL, };
+
+
+// i/o
+
+struct Pos c_println_Int(const Int n) {
     printf("%ld\n", n);
+    return Unit;
 }
 
-void c_println_pos_boolean(const struct Pos p) {
+struct Pos c_println_Boolean(const struct Pos p) {
     printf("%s\n", p.tag ? "true" : "false");
+    return Unit;
 }
 
+struct Pos c_println_Double(const Double x) {
+    printf("%g\n", x);
+    return Unit;
+}
+
+
+// arithmetic
+
+#define C_OP(T, OP_NAME, OP) \
```

Comment:
Nice idea of sharing code, but please hard-code the operations in `llvm/effekt.effekt` instead. This will lead to less code and less indirection.